### PR TITLE
Improve OntologyAnnotation equality

### LINKED ISSUE #268: [BUG] Improve OntologyAnnotation Equality and Hash
Might give wrong results when ShortAnnotation cannot be retrieved.

### PR DESCRIPTION
**Pull request by Freymaurer**
Improve OntologyAnnotation hash and equality handling. Added tests. 

Closes #268

### CODE REVIEW

File: ARCtrl.sln
Comment (HLWeil) [inline ARCtrl.sln:1]: Maybe mention as alternative in README?

File: setup.cmd [hunk 1]
```diff
@@ -0,0 +1,15 @@
+@echo off
```
Comment (HLWeil) [inline setup.cmd:1]: Does `@echo off` potentially obscure some useful information like "what command is called"?

File: src/Core/OntologyAnnotation.fs [hunk 2]
```diff
@@ -131,31 +131,28 @@ type OntologyAnnotation(?name,?tsr,?tan, ?comments) =
...
-                other.TermAccessionShort = this.TermAccessionShort
-                ||
-                other.TermAccessionOntobeeUrl = this.TermAccessionOntobeeUrl
-            elif this.Name.IsSome && other.Name.IsSome then
-                other.NameText = this.NameText
-            elif this.TermAccessionNumber.IsNone && other.TermAccessionNumber.IsNone && this.Name.IsNone && other.Name.IsNone then
-                true
-            else 
-                false
+    override this.GetHashCode() =
```
Comment (HLWeil) [inline src/Core/OntologyAnnotation.fs:134]: How does this hash behave in cases, where the underlying Term is the same but once stored as obo url and once as short tan?

File: src/Core/OntologyAnnotation.fs [hunk 3]
```diff
@@ -131,31 +131,28 @@ type OntologyAnnotation(?name,?tsr,?tan, ?comments) =
...
-            elif this.TermAccessionNumber.IsNone && other.TermAccessionNumber.IsNone && this.Name.IsNone && other.Name.IsNone then
-                true
-            else 
-                false
+    override this.GetHashCode() = 
+        [|
+            HashCodes.boxHashOption this.Name
+            match this.TermSourceREF, this.TANInfo with
+            | None, Some taninfo -> // if we get taninfo we assume tsr to be inferrable by taninfo
+                HashCodes.hash {|tsr = taninfo.IDSpace; tan = taninfo.IDSpace + ":" + taninfo.LocalID|}
```
Comment (HLWeil) [inline src/Core/OntologyAnnotation.fs:139]: I would make this explicit, i.e. call hashes on the fields and merge them. Record Type hashes behave a bit weird for fable Python

File: src/Core/Process/Factor.fs [hunk 4]
```diff
@@ -32,7 +32,7 @@ type Factor =
 
     /// Get ISATab string entries from an ISAJson Factor object
     static member toString (factor : Factor) =
```
Comment (HLWeil) [inline src/Core/Process/Factor.fs:34]: Should this also be renamed to "toStringOption"

File: src/Core/Process/MaterialAttribute.fs [hunk 5]
```diff
@@ -29,7 +29,7 @@ type MaterialAttribute =
 
     /// Get ISATab string entries from an ISAJson MaterialAttribute object
     static member toString (ma : MaterialAttribute) =
```
Comment (HLWeil) [inline src/Core/Process/MaterialAttribute.fs:31]: Should this also be renamed to "toStringOption"?

File: src/Core/Process/ProtocolParameter.fs [hunk 6]
```diff
@@ -31,7 +31,7 @@ type ProtocolParameter =
 
     /// Get ISATab string entries from an ISAJson ProtocolParameter object (name,source,accession)
     static member toString (pp : ProtocolParameter) =
```
Comment (HLWeil) [inline src/Core/Process/ProtocolParameter.fs:33]: Should this also be renamed to "toStringOption"

File: tests/Core/OntologyAnnotation.Tests.fs [hunk 7]
```diff
@@ -0,0 +1,117 @@
+﻿module OntologyAnnotation.Tests
```
Comment (HLWeil) [inline tests/Core/OntologyAnnotation.Tests.fs:1]: Did you do changes besides moving?

**Review by HLWeil — CHANGES_REQUESTED** [review]

File: ARCtrl.sln
Comment (Freymaurer) [reply ARCtrl.sln:1]: mention what? 😅

File: setup.cmd (hunk 1 above)
Comment (Freymaurer) [reply setup.cmd:1]: At least not for me

File: src/Core/OntologyAnnotation.fs (hunk 2 above)
Comment (Freymaurer) [reply src/Core/OntologyAnnotation.fs:134]: taninfo uses information after regex parsing. So everything we support with our regex works fine. I also added some tests for this.

File: src/Core/OntologyAnnotation.fs (hunk 3 above)
Comment (Freymaurer) [reply src/Core/OntologyAnnotation.fs:139]: Yeah really not sure about this. I tried to remove the python specific code and it returned strange numbers. But i must check tomorrow in more detail

File: src/Core/Process/Factor.fs (hunk 4 above)
Comment (Freymaurer) [reply src/Core/Process/Factor.fs:34]: I think i named it "toStringObject", but yes we should rename. Not sure to what.

File: tests/Core/OntologyAnnotation.Tests.fs (hunk 7 above)
Comment (Freymaurer) [reply tests/Core/OntologyAnnotation.Tests.fs:1]: i added all "equal" tests

File: ARCtrl.sln
Comment (HLWeil) [reply ARCtrl.sln:1]: Whoops seems like the comment was crooked :D

I meant the `setup.cmd`

Comment (Freymaurer) [reply ARCtrl.sln:1]: Yes can do!

**Review by HLWeil — APPROVED** [review]

**Pull request merged**